MobiFlight/Joysticks/VKB/VKBDevice.cs

```diff
@@ -0,0 +1,192 @@
+            Definition = definition;
+            if (Device == null)
+            {
+                Device = GetMatchingHidDevice(joystick);
+                if (Device == null) return;
+            }
+        }
+        public override void Connect(IntPtr handle)
+        {
+            base.Connect(handle);
+            if (Stream == null)
+            {
+                Stream = Device.Open();
+                Stream.ReadTimeout = System.Threading.Timeout.Infinite;
+            }
+
+            if (InputReceiver == null)
+            {
+                InputReceiver = new ReportDescriptor(VKBHidReport.Descriptor).CreateHidDeviceInputReceiver();
+                // We use our own descriptor since the one we get from Windows/HIDSharp has been altered and encoder data is missing
+                InputReceiver.Received += OnHidReportReceived;
+                InputReceiver.Start(Stream);
+            }
+        }
+        protected override void SendData(byte[] data)
+        {
+            // VKBLedContainer has its own handling that replaces RequiresOutputUpdate, so we just send.
+            // Don't try and send data if no outputs are defined.
+            if (Definition?.Outputs == null || Definition?.Outputs.Count == 0)
+            {
+                return;
+            }
+            Stream?.SetFeature(data);
+
+        }
+        protected override void EnumerateDevices()
+        {
+            base.EnumerateDevices();
+            SortedList<byte, JoystickDevice> EncoderDecList = new SortedList<byte, JoystickDevice>();
+            SortedList<byte, JoystickDevice> EncoderIncList = new SortedList<byte, JoystickDevice>();
+            if (Definition == null || Definition.Inputs == null) return;
+            foreach (var input in Definition.Inputs)
+            {
+                // The 1xxx range is limited to encoders. They are also not fed from DirectInput.
+                if (input.Id >= 1000 && input.Id < 2000 && input.Type == JoystickDeviceType.Button)
+                {
+                    byte encoderIndex = GetEncoderIndex(input);
+                    VKBEncoder.EncoderAction encoderAction = GetEncoderAction(input);
+                    if (encoderAction == VKBEncoder.EncoderAction.DEC)
+                    {
+                        EncoderDecList.Add(encoderIndex, new JoystickDevice { Name = $"Button {1000 + 10 * encoderIndex + (int)encoderAction}", Label = input.Label, Type = DeviceType.Button, JoystickDeviceType = input.Type });
+                    }
+                    if (encoderAction == VKBEncoder.EncoderAction.INC)
+                    {
+                        EncoderIncList.Add(encoderIndex, new JoystickDevice { Name = $"Button {1000 + 10 * encoderIndex + (int)encoderAction}", Label = input.Label, Type = DeviceType.Button, JoystickDeviceType = input.Type });
+                    }
+                    Buttons.FindAll(but => but.Label == input.Label).ForEach(but => but.Label += " (Legacy DirectInput)"); 
```

Comment:
Will users know what "(Legacy DirectInput)" means? I assume that will eventually get shown in the UI. Is it necessary?

Reply:
It is necessary to put *something* there. Removing it entirely would require a new copy of the button handling code that can handle gaps in the button list, plus it would be bad for users who have reconfigured their device to reuse those button slots for different purposes, or for those whose devices are not set up to use the encoder API (old firmware or wrong firmware settings).
Basically, the purpose is to have something that tells users "try the other one first, but if that does not work this option still might". I am open to suggestions how this could be presented in a way that is more accessible to users.